File: inst/gha/validate-test-files.R
# Execute with `source("inst/gha/validate-test-files.R")`

errors_found <- list()

app_folders <- basename(list.dirs("inst/apps", recursive = FALSE))
app_folder_nums <- sub("^(\\d\\d\\d)-.*$", "\\1", app_folders)
# Can not call `setdiff()`. This internally calls `unique()`
app_folder_nums <- app_folder_nums[!(app_folder_nums %in% c("000"))]
if (any(duplicated(app_folder_nums))) {
  stop("Duplicate app numbers found: ", paste0(app_folder_nums[duplicated(app_folder_nums)], collapse = ", "))
}

for (app_path in list.dirs("inst/apps", recursive = FALSE)) {
  tryCatch({

    app_files <- dir(app_path, pattern = "\\.(R|Rmd)$", full.names = TRUE)
    tests_path <- file.path(app_path, "tests")
    if (dir.exists(tests_path)) {
      runners <- dir(tests_path, pattern = "R$")
      if (length(runners) > 1) {
        stop("More than 1 test runner found in ", app_path, ". Found: ", paste0(runners, collapse = ", "))
      }
      # Verify simple testthat.R
      testthat_path <- file.path(tests_path, "testthat.R")
      if (!file.exists(testthat_path)) {
        stop("Missing `testthat.R` for app: ", app_path)
      }
      testthat_lines <- readLines(testthat_path)
      if (length(testthat_lines) > 1) {
        stop("Non-basic testthat script found for ", testthat_path, ". Found:\n", paste0(testthat_lines, "\n"))
      }
      if (testthat_lines != "shinytest2::test_app()") {
        stop("Non-shinytest2 testthat script found for ", testthat_path, ". Found:\n", paste0(testthat_lines, "\n"))
      }

      # Verify shinyjster content
      shinyjster_file <- file.path(tests_path, "testthat", "test-shinyjster.R")
      if (file.exists(shinyjster_file)) {
        for (jster_txt in c("shinyjster_server(", "shinyjster_js(")) {
          found <- FALSE
          for (app_file in app_files) {
            if (any(grepl(jster_txt, readLines(app_file), fixed = TRUE))) {
              found <- TRUE
              break
            }
          }
          if (!found) {
            stop(app_path, " did not contain ", jster_txt, " but contains a `./tests/testthat/test-shinyjster.R")
          }
        }
      }

    } else {
      # Test for manual app
      found <- FALSE
      for (app_file in app_files) {
        if (any(grepl("shinycoreci::::is_manual_app", readLines(app_file), fixed = TRUE))) {
          found <- TRUE
          break
        }
      }
      if (!found) {
        stop(
          "No `./", app_path, "/tests` folder found for non-manual app.\n",
          "Either add tests with `shinytest2::use_shinytest2('", app_path, "')`\n",
          "Or set to manual by calling `shinycoreci::use_manual_app('", app_path, "')`"
        )
      }
    }

    # # Make sure shinycoreci is not used within an app
    # for (file in dir(app_path, recursive = TRUE, full.names = TRUE, pattern = "\\.(R|Rmd)$")) {
    #   # Ignore first 000 apps
    #   if (grepl("^inst/apps/000-", file)) next

    #   file_lines <- readLines(file)
    #   if (any(grepl("shinycoreci)", file_lines, fixed = TRUE))) {
    #     stop("File `", file, "` contains library() or require() call to {shinycoreci}. Remove usage of {shinycoreci}.")
    #   }
    #   file_lines <- gsub("shinycoreci::::", "shinycoreci____", file_lines)
    #   if (any(grepl("shinycoreci::", file_lines, fixed = TRUE))) {
    #     stop("File `", file, "` contains usage of {shinycoreci}. Replace this code.")
    #   }
    # }
  }, error = function(e) {
    errors_found[[length(errors_found) + 1]] <<- e
  })
}

if (length(errors_found) > 0) {
  for (e in errors_found) {
    message("\n", e)
  }
  stop("Errors found when validating apps")
}

# warns <- warnings()
# if (length(warns) > 0) {
#   warn_txt <- Map(names(warns), warns, f = function(msg, expr) { paste0(msg, " : ", as.character(as.expression(expr)), "\n") })
#   stop("Warnings found when validating apps:\n", paste0(warn_txt, collapse = ""))
# }

message("No errors found when validating apps")

File: inst/gha/validate-test-files.py
# Execute with `python inst/gha/validate-test-files.py`

import os
import re
import glob
import sys
from pathlib import Path

# Set the working directory to the repo root
here = Path(__file__).resolve().parent.parent.parent
os.chdir(here)


def get_app_folders():
    app_folders = [
        f for f in glob.glob("inst/apps/*", root_dir=here) if os.path.isdir(f)
    ]
    app_folders.sort()
    app_folder_nums = [
        re.sub(r"^(\d\d\d)-.*$", r"\1", os.path.basename(folder))
        for folder in app_folders
    ]
    # Filter out "000" folders
    app_folder_nums = [num for num in app_folder_nums if num != "000"]
    # Check for duplicates
    duplicates = set([x for x in app_folder_nums if app_folder_nums.count(x) > 1])
    if duplicates:
        raise Exception(f"Duplicate app numbers found: {', '.join(duplicates)}")
    return app_folders


def validate_app(app_path):
    print(f"Validating app: {app_path}")
    app_files = glob.glob(os.path.join(app_path, "*.R")) + glob.glob(
        os.path.join(app_path, "*.Rmd")
    )
    tests_path = os.path.join(app_path, "tests")

    if os.path.isdir(tests_path):
        runners = [f for f in os.listdir(tests_path) if f.endswith("R")]
        if len(runners) > 1:
            raise Exception(
                f"More than 1 test runner found in {app_path}. Found: {', '.join(runners)}"
            )

        # Verify simple testthat.R
        testthat_path = os.path.join(tests_path, "testthat.R")
        if not os.path.exists(testthat_path):
            raise Exception(f"Missing `testthat.R` for app: {app_path}")

        with open(testthat_path, "r") as f:
            testthat_lines = f.readlines()
            testthat_lines = [line.strip() for line in testthat_lines if line.strip()]

        if len(testthat_lines) > 1:
            raise Exception(
                f"Non-basic testthat script found for {testthat_path}. Found:\n{''.join(testthat_lines)}"
            )

        if testthat_lines[0] != "shinytest2::test_app()":
            raise Exception(
                f"Non-shinytest2 testthat script found for {testthat_path}. Found:\n{''.join(testthat_lines)}"
            )

        # Verify shinyjster content
        shinyjster_file = os.path.join(tests_path, "testthat", "test-shinyjster.R")
        if os.path.exists(shinyjster_file):
            for jster_txt in ["shinyjster_server(", "shinyjster_js("]:
                found = False
                for app_file in app_files:
                    with open(app_file, "r") as f:
                        if any(jster_txt in line for line in f):
                            found = True
                            break
                if not found:
                    raise Exception(
                        f"{app_path} did not contain {jster_txt} but contains a `./tests/testthat/test-shinyjster.R"
                    )
    else:
        # Test for manual app
        found = False
        for app_file in app_files:
            with open(app_file, "r") as f:
                if any("shinycoreci::::is_manual_app" in line for line in f):
                    found = True
                    break
        if not found:
            raise Exception(
                f"No `./{app_path}/tests` folder found for non-manual app.\n"
                f"Either add tests with `shinytest2::use_shinytest2('{app_path}')`\n"
                f"Or set to manual by calling `shinycoreci::use_manual_app('{app_path}')`"
            )

    # The commented-out code for checking shinycoreci usage would go here
    # Uncomment if needed and adapt the R code accordingly


def main():
    app_folders = get_app_folders()
    # print(f"Found {len(app_folders)} app folders: {', '.join(app_folders)}")

    errors_found = []
    for app_folder in app_folders:
        try:
            validate_app(app_folder)
        except Exception as e:
            errors_found.append(e)

    if errors_found:
        for e in errors_found:
            print(f"\n{e}")
        sys.exit("Errors found when validating apps")
    else:
        print("No errors found when validating apps")


if __name__ == "__main__":
    main()
